sql/moz-fx-data-shared-prod/firefox_accounts_derived/fxa_users_services_first_seen_v1/query.sql

```diff
@@ -1,136 +1,93 @@
       ORDER BY
         `timestamp`
+      ASC
     )
-),
-  -- we need this next section because `did_register` will be BOTH true and false within the flows that the user registered on.
-  -- this dedupes the rows from above and sets did_register to true only on flows that included a registration
-  -- I've verified that `date(first_service_timestamp), count(distinct user_id) where did_register = true group by 1`  matches the counts of registrations per day in amplitude.
-first_services_g AS (
-  SELECT
-    * EXCEPT (_n)
-  FROM
-    first_services
-  WHERE
-    _n = 1
-),
-  -- sadly, `entrypoint` is null on registration complete and login complete events.
-  -- this means we have to use first_service_flow to join back on the original source table's flow_id,
-  -- and take the first occurrence of `entrypoint` within the flow that the user first appeared in the service on.
-flows AS (
-  SELECT DISTINCT
-    s.first_service_flow,
-    FIRST_VALUE(f.entrypoint) OVER (
-      PARTITION BY
-        f.flow_id
-      ORDER BY
-        f.`timestamp`
-      ROWS BETWEEN
-        UNBOUNDED PRECEDING
-        AND UNBOUNDED FOLLOWING
-    ) AS first_service_entrypoint,
-    FIRST_VALUE(f.timestamp) OVER (
-      PARTITION BY
-        f.flow_id
-      ORDER BY
-        f.`timestamp`
-      ROWS BETWEEN
-        UNBOUNDED PRECEDING
-        AND UNBOUNDED FOLLOWING
-    ) AS first_service_timestamp,
-    FIRST_VALUE(f.country) OVER (
-      PARTITION BY
-        f.flow_id
-      ORDER BY
-        f.`timestamp`
-      ROWS BETWEEN
-        UNBOUNDED PRECEDING
-        AND UNBOUNDED FOLLOWING
-    ) AS first_service_country,
-    FIRST_VALUE(f.os_name) OVER (
-      PARTITION BY
-        f.flow_id
-      ORDER BY
-        f.`timestamp`
-      ROWS BETWEEN
-        UNBOUNDED PRECEDING
-        AND UNBOUNDED FOLLOWING
-    ) AS first_service_os
-  FROM
-    first_services_g s
-  INNER JOIN
-    firefox_accounts.fxa_content_auth_oauth_events AS f
-  ON
-    s.first_service_flow = f.flow_id
-  WHERE
-    f.entrypoint IS NOT NULL
-    AND s.first_service_flow IS NOT NULL
-    AND DATE(f.`timestamp`) >= '2019-03-01'
 )
-  -- finally take the entrypoint data and join it back on the other information (os, country etc).
-  -- also, add a row number that indicates the order in which the user signed up for their services.
+
 SELECT
-  s.user_id,
-  s.service,
-  s.first_service_flow,
-  s.did_register,
-  f.first_service_entrypoint AS entrypoint,
-  f.first_service_timestamp,
-  f.first_service_country,
-  f.first_service_os,
-  ROW_NUMBER() OVER (PARTITION BY s.user_id ORDER BY first_service_timestamp_last) AS service_number
-FROM
-  first_services_g s
-LEFT JOIN
-  flows f
-USING
-  (first_service_flow)
-WHERE
-  first_service_flow IS NOT NULL
+  * EXCEPT(row_id)
+FROM (
+  SELECT
+    *,
+    ROW_NUMBER() OVER(PARTITION BY user_id, service, os_name ORDER BY first_seen_date ASC) AS row_id
+  FROM first_values
+  WHERE user_id IS NOT NULL
+)
+WHERE row_id = 1
```

Comment:
this can be done without a subquery using `QUALIFY`:
```suggestion
  *
FROM first_values
WHERE user_id IS NOT NULL
QUALIFY
    ROW_NUMBER() OVER(PARTITION BY user_id, service, os_name ORDER BY first_seen_date ASC) = 1
```

Reply:
This is indeed much cleaner, thanks for the suggestion 🙇